synapse/models/files.py

```diff
@@ -1,107 +1,156 @@
+            raise s_exc.BadTypeValu(name=self.name, valu=valu, mesg=mesg)
+
+        subs = {}
+        if norm.find('.') != -1:
+            subs['ext'] = norm.rsplit('.', 1)[1]
+
+        return norm, {'subs': subs}
+
+    def indx(self, norm):
+        return norm.encode('utf8')
+
+class FilePath(s_types.Type):
+
+    def postTypeInit(self):
+        self.setNormFunc(str, self._normPyStr)
+        self.indxcmpr['^='] = self.indxByPref
+
+    def indxByPref(self, valu):
+        valu = value.strip().lower().replace('\\', '/')
+        indx = valu.encode('utf8')
+        return (
+            ('pref', indx),
+        )
+
+    def _normPyStr(self, valu):
+
+        lead = ''
+        if valu[0] == '/':
+            lead = '/'
+
+        valu = valu.strip().lower().replace('\\', '/').strip('/')
+        if not valu:
+            return ''
+
+        path = []
+
+        for part in valu.split('/'):
+
+            if part == '.':
+                continue
+
+            if part == '..':
+                if len(path):
+                    path.pop()
+
+                continue
+
+            path.append(part)
+
+        fullpath = lead + '/'.join(path)
+
+        subs = {'base': path[-1]}
+        if len(path) > 1:
+            subs['dir'] = lead + '/'.join(path[:-1])
+
+        return fullpath, {'subs': subs}
+
+    def indx(self, norm):
+        return norm.encode('utf8')
 
 class FileBytes(s_types.Type):
 
     def postTypeInit(self):
         self.setNormFunc(str, self._normPyStr)
         self.setNormFunc(bytes, self._normPyBytes)
 
-    def indx(self, valu):
-        return s_common.uhex(valu)
+    def indx(self, norm):
+        return norm.encode('utf8')
 
     def _normPyStr(self, valu):
 
-        if len(valu) != 64:
+        if valu == '*':
+            guid = s_common.guid()
+            norm = f'guid:{guid}'
+            return norm, {}
+
+        if valu.find(':') == -1:
+
+            # we're ok with un-adorned sha256s
+            if len(valu) == 64 and s_common.uhex(valu):
+                valu = valu.lower()
+                return f'sha256:{valu}', {}
+
             raise s_exc.BadTypeValu(name=self.name, valu=valu)
 
-        s_common.uhex(valu)
+        kind, valu = valu.split(':')
+
+        if kind == 'base64':
+            byts = base64.b64decode(valu)
+            return self._normPyBytes(byts)
 
-        norm = valu.lower()
-        return norm, {}
+        valu = valu.lower()
+
+        if kind == 'hex':
+            byts = s_common.uhex(valu)
+            return self._normPyBytes(byts)
+
+        if kind == 'guid':
+
+            valu = valu.lower()
+            if not s_common.isguid(valu):
+                raise s_exc.BadTypeValu(name=self.name, valu=valu)
+
+            return f'guid:{valu}', {}
+
+        if kind == 'sha256':
+
+            if len(valu) != 64:
+                raise s_exc.BadTypeValu(name=self.name, valu=valu)
+
+            s_common.uhex(valu)
+            return f'sha256:{valu}', {}
```

Comment:
if we are assuming that the `sha256:` is a sha256, we should kick out a `sha:256` sub

Reply:
Oh, right.  Duh.  Thx.